File: frontend/src/services/gameStorage.ts
// Game Storage Service - Save/Resume functionality
import AsyncStorage from '@react-native-async-storage/async-storage';
import { CrosswordPuzzle, GameSettings } from '../types';

const STORAGE_KEYS = {
  SAVED_GAME: '@crossword_saved_game',
  GAME_STATS: '@crossword_game_stats',
};

export interface SavedGame {
  puzzle: CrosswordPuzzle;
  settings: GameSettings;
  elapsedTime: number;
  savedAt: number;
}

export interface GameStats {
  gamesPlayed: number;
  gamesCompleted: number;
  totalTime: number;
  bestTime: number | null;
  lastPlayed: number | null;
}

// Save current game
export async function saveGame(
  puzzle: CrosswordPuzzle,
  settings: GameSettings,
  elapsedTime: number
): Promise<void> {
  try {
    const savedGame: SavedGame = {
      puzzle,
      settings,
      elapsedTime,
      savedAt: Date.now(),
    };
    await AsyncStorage.setItem(STORAGE_KEYS.SAVED_GAME, JSON.stringify(savedGame));
  } catch (error) {
    console.error('Error saving game:', error);
  }
}

// Load saved game
export async function loadSavedGame(): Promise<SavedGame | null> {
  try {
    const savedGameStr = await AsyncStorage.getItem(STORAGE_KEYS.SAVED_GAME);
    if (savedGameStr) {
      return JSON.parse(savedGameStr) as SavedGame;
    }
    return null;
  } catch (error) {
    console.error('Error loading saved game:', error);
    return null;
  }
}

// Delete saved game
export async function deleteSavedGame(): Promise<void> {
  try {
    await AsyncStorage.removeItem(STORAGE_KEYS.SAVED_GAME);
  } catch (error) {
    console.error('Error deleting saved game:', error);
  }
}

// Check if there's a saved game
export async function hasSavedGame(): Promise<boolean> {
  try {
    const savedGameStr = await AsyncStorage.getItem(STORAGE_KEYS.SAVED_GAME);
    return savedGameStr !== null;
  } catch (error) {
    console.error('Error checking saved game:', error);
    return false;
  }
}

// Load game stats
export async function loadGameStats(): Promise<GameStats> {
  try {
    const statsStr = await AsyncStorage.getItem(STORAGE_KEYS.GAME_STATS);
    if (statsStr) {
      return JSON.parse(statsStr) as GameStats;
    }
    return {
      gamesPlayed: 0,
      gamesCompleted: 0,
      totalTime: 0,
      bestTime: null,
      lastPlayed: null,
    };
  } catch (error) {
    console.error('Error loading game stats:', error);
    return {
      gamesPlayed: 0,
      gamesCompleted: 0,
      totalTime: 0,
      bestTime: null,
      lastPlayed: null,
    };
  }
}

// Update game stats
export async function updateGameStats(
  completed: boolean,
  time: number
): Promise<void> {
  try {
    const stats = await loadGameStats();
    
    stats.gamesPlayed++;
    stats.totalTime += time;
    stats.lastPlayed = Date.now();
    
    if (completed) {
      stats.gamesCompleted++;
      if (stats.bestTime === null || time < stats.bestTime) {
        stats.bestTime = time;
      }
    }
    
    await AsyncStorage.setItem(STORAGE_KEYS.GAME_STATS, JSON.stringify(stats));
  } catch (error) {
    console.error('Error updating game stats:', error);
  }
}

// Format time for display
export function formatTime(seconds: number): string {
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${mins.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
}

// Format date for display
export function formatDate(timestamp: number): string {
  const date = new Date(timestamp);
  const day = date.getDate().toString().padStart(2, '0');
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const hours = date.getHours().toString().padStart(2, '0');
  const minutes = date.getMinutes().toString().padStart(2, '0');
  return `${day}.${month} ${hours}:${minutes}`;
}

File: android/app/src/main/java/com/bulgarianwordgrid/services/GameStorage.kt
package com.bulgarianwordgrid.services

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import androidx.datastore.preferences.preferencesDataStore
import com.bulgarianwordgrid.model.*
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

internal val Context.gameDataStore: DataStore<Preferences> by preferencesDataStore(name = "crossword_prefs")

object GameStorage {
    private val SAVED_GAME_KEY = stringPreferencesKey("saved_game")
    private val GAME_STATS_KEY = stringPreferencesKey("game_stats")

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun saveGame(context: Context, puzzle: CrosswordPuzzle, settings: GameSettings, elapsedTime: Int) {
        try {
            val savedGame = SavedGame(
                puzzle = puzzle,
                settings = settings,
                elapsedTime = elapsedTime,
                savedAt = System.currentTimeMillis()
            )
            context.gameDataStore.edit { prefs ->
                prefs[SAVED_GAME_KEY] = json.encodeToString(savedGame)
            }
        } catch (e: Exception) {
            android.util.Log.e("GameStorage", "Error saving game", e)
        }
    }

    suspend fun loadSavedGame(context: Context): SavedGame? {
        return try {
            val prefs = context.gameDataStore.data.first()
            val savedGameStr = prefs[SAVED_GAME_KEY]
            if (savedGameStr != null) json.decodeFromString<SavedGame>(savedGameStr) else null
        } catch (e: Exception) {
            android.util.Log.e("GameStorage", "Error loading saved game", e)
            null
        }
    }

    suspend fun deleteSavedGame(context: Context) {
        try {
            context.gameDataStore.edit { it.remove(SAVED_GAME_KEY) }
        } catch (e: Exception) {
            android.util.Log.e("GameStorage", "Error deleting saved game", e)
        }
    }

    suspend fun hasSavedGame(context: Context): Boolean {
        return try {
            val prefs = context.gameDataStore.data.first()
            prefs[SAVED_GAME_KEY] != null
        } catch (e: Exception) {
            false
        }
    }

    suspend fun loadGameStats(context: Context): GameStats {
        return try {
            val prefs = context.gameDataStore.data.first()
            val statsStr = prefs[GAME_STATS_KEY]
            if (statsStr != null) json.decodeFromString<GameStats>(statsStr) else GameStats()
        } catch (e: Exception) {
            android.util.Log.e("GameStorage", "Error loading game stats", e)
            GameStats()
        }
    }

    suspend fun updateGameStats(context: Context, completed: Boolean, time: Int) {
        try {
            val stats = loadGameStats(context)
            val updated = stats.copy(
                gamesPlayed = stats.gamesPlayed + 1,
                totalTime = stats.totalTime + time,
                lastPlayed = System.currentTimeMillis(),
                gamesCompleted = if (completed) stats.gamesCompleted + 1 else stats.gamesCompleted,
                bestTime = if (completed) {
                    if (stats.bestTime == null || time < stats.bestTime) time else stats.bestTime
                } else stats.bestTime
            )
            context.gameDataStore.edit { prefs ->
                prefs[GAME_STATS_KEY] = json.encodeToString(updated)
            }
        } catch (e: Exception) {
            android.util.Log.e("GameStorage", "Error updating game stats", e)
        }
    }

    fun formatTime(seconds: Int): String {
        val mins = seconds / 60
        val secs = seconds % 60
        return "%02d:%02d".format(mins, secs)
    }

    fun formatDate(timestamp: Long): String {
        val cal = java.util.Calendar.getInstance()
        cal.timeInMillis = timestamp
        val day = "%02d".format(cal.get(java.util.Calendar.DAY_OF_MONTH))
        val month = "%02d".format(cal.get(java.util.Calendar.MONTH) + 1)
        val hours = "%02d".format(cal.get(java.util.Calendar.HOUR_OF_DAY))
        val minutes = "%02d".format(cal.get(java.util.Calendar.MINUTE))
        return "$day.$month $hours:$minutes"
    }
}
